addons/sale/models/payment_transaction.py

```diff
@@ -49,34 +46,38 @@ def _set_pending(self, state_message=None):
             sales_orders._send_order_confirmation_mail()
 
     def _check_amount_and_confirm_order(self):
-        self.ensure_one()
-        for order in self.sale_order_ids.filtered(lambda so: so.state in ('draft', 'sent')):
-            if order.currency_id.compare_amounts(self.amount, order.amount_total) == 0:
-                order.with_context(send_email=True).action_confirm()
-            else:
-                _logger.warning(
-                    '<%s> transaction AMOUNT MISMATCH for order %s (ID %s): expected %r, got %r',
-                    self.acquirer_id.provider,order.name, order.id,
-                    order.amount_total, self.amount,
-                )
-                order.message_post(
-                    subject=_("Amount Mismatch (%s)", self.acquirer_id.provider),
-                    body=_("The order was not confirmed despite response from the acquirer (%s): order total is %r but acquirer replied with %r.") % (
-                        self.acquirer_id.provider,
-                        order.amount_total,
-                        self.amount,
+        """ Confirm the sale order based on the amount of a transaction.
+
+        :return: confirmed sale orders
+        :rtype: a `sale.order` recordset
+        """
+        confirmed_orders = self.env['sale.order']
+        for tx in self:
+            # We only support the flow where exactly one quotation is linked to a transaction and
+            # vice versa.
+            order = tx.sale_order_ids.filtered(lambda so: so.state in ('draft', 'sent'))
+            if len(tx.sale_order_ids) == 1 and order and len(order.transaction_ids) == 1:
+                # Check if the SO is fully paid
+                if order.currency_id.compare_amounts(tx.amount, order.amount_total) == 0:
+                    order.with_context(send_email=True).action_confirm()
+                    confirmed_orders |= order
+                else:
+                    _logger.warning(
+                        '<%s> transaction AMOUNT MISMATCH for order %s (ID %s): expected %s, got %s',
+                        tx.provider,
+                        order.name,
+                        order.id,
+                        format_amount(order.env, order.amount_total, order.currency_id),
+                        format_amount(tx.env, tx.amount, tx.currency_id),
```

Comment:
That's %s a %s lot %s of %s parameters. It'd be clearer if you use a `dict` for the interpolation.